### PR TITLE
Evenly spread queriers across available nodes

### PR DESCRIPTION
**Pull request by salvacorts**
**What this PR does / why we need it**:

Currently, queriers run on different nodes. We do this by using the following anti-affinity rule:

```yaml
affinity:
  podAntiAffinity:
    requiredDuringSchedulingIgnoredDuringExecution:
    - labelSelector:
        matchLabels:
          name: querier
      topologyKey: kubernetes.io/hostname
```

The original reasoning behind this is to make our systems more resilient to node failures. For example, if we have 5 queriers, all scheduled in the same node, and the node fails, we would not be able to serve queries until the pods get rescheduled and running on another node.

When new queriers are created, they need to be scheduled into different nodes. If there are not enough nodes in the cluster, new nodes need to be provisioned. This has two implications:
- Cost: Each node has an implicit cost on top of the workload we deploy into the node.
- Latency: It takes longer to allocate a new node into the cluster, and then schedule the pod, than to schedule it right away into an already available node.

This PR uses Kubernetes's `TopologySpreadConstraints` to evenly spread queriers across the available nodes, but allows more than one querier to be scheduled into the same node.

**Checklist**
- [ ] Documentation added
- [ ] Tests updated
- [x] Is this an important fix or new feature? Add an entry in the `CHANGELOG.md`.
- [x] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

**Comment by salvacorts** [comment]
As suggested by @dannykopping in the design doc, I added an entry to the changelog and the upgrade guide.

**Review by dannykopping — COMMENTED** [review]
This looks good @salvacorts, but I think we should make it configurable.

> If you want to keep running up to one querier per node, you will need to revert the changes for `production/ksonnet/loki/querier.libsonnet`
made at [6415](https://github.com/grafana/loki/pull/6415).

How about we offer both the node affinity & the topology spread via a config value (defaulted to topology spread), and allow configuring the max skew?

This could later be expanded for all components (defaulted to node affinity for now)

**Comment by salvacorts** [comment]
> How about we offer both the node affinity & the topology spread via a config value (defaulted to topology spread), and allow configuring the max skew?

@dannykopping That's a good idea. I just added two new config parameters: `querier.use_topology_spread` and `querier.topology_spread_max_skew`. I also updated the upgrade guide to reflect these changes.

File: docs/sources/upgrading/_index.md [hunk 1]
```diff
@@ -36,8 +36,7 @@ The output is incredibly verbose as it shows the entire internal config struct u
 #### Evenly spread queriers across kubernetes nodes
 
 We now evenly spread queriers across the available kubernetes nodes, but allowing more than one querier to be scheduled into the same node.
-If you want to keep running up to one querier per node, you will need to revert the changes for `production/ksonnet/loki/querier.libsonnet`
-made at [6415](https://github.com/grafana/loki/pull/6415).
+If you want to keep running up to one querier per node, set `$._config.querier.use_topology_spread` to false.
```
Comment (dannykopping) [inline docs/sources/upgrading/_index.md:39]:
```suggestion
If you want to run at most a single querier per node, set `$._config.querier.use_topology_spread` to false.
```

File: production/ksonnet/loki/config.libsonnet [hunk 2]
```diff
@@ -47,6 +47,13 @@
       // A higher value will lead to a querier trying to process more requests than there are available
       // cores and will result in scheduling delays.
       concurrency: 4,
+
+      // If use_topology_spread is true, queriers can run on nodes already running queriers but will be
+      // spread through the available nodes using a TopologySpreadConstraints with a max skew
```
Comment (dannykopping) [inline production/ksonnet/loki/config.libsonnet:52]: Can we link the docs here pls?

**Review by dannykopping — APPROVED** [review]
LGTM! Two minor nits

File: production/ksonnet/loki/config.libsonnet (hunk 2 above)
Comment (salvacorts) [reply production/ksonnet/loki/config.libsonnet:52]: Done

File: docs/sources/upgrading/_index.md (hunk 1 above)
Comment (salvacorts) [reply docs/sources/upgrading/_index.md]: Done

**Pull request merged**